### PR TITLE
Allow laptops to include optional USB keyboard support

### PR DESCRIPTION
**Pull request by JonathonHall-Purism**
Laptops can include optional USB keyboard support (default off unless the board also sets the default to 'y').  The setting is in the configuration GUI.

CONFIG_USER_USB_KEYBOARD is now the user-controlled setting on those boards.  'CONFIG_USB_KEYBOARD' is no longer used to avoid any conflict with prior releases that expect this to be a compile-time setting only (conflicts risk total lock out requiring hardware flash, so some caution is justified IMO).

Boards previously exporting CONFIG_USB_KEYBOARD now export CONFIG_USB_KEYBOARD_REQUIRED.  Those boards don't have built-in keyboards, USB keyboard is always enabled. (librem_mini, librem_mini_v2, librem_11, librem_l1um, librem_l1um_v2, talos-2, kgpe-d16_workstation-usb_keyboard, x230-hotp-maximized_usb-kb).

Librem laptops now export CONFIG_SUPPORT_USB_KEYBOARD to enable optional support.  The default is still 'off'.

### CODE REVIEW

**Comment by JonathonHall-Purism** [comment]
This was requested by a few PureBoot users BTW.  I think it still has to be off by default on laptops to avoid reducing any security guarantees offered by Qubes, but users can choose to enable it now on boards including the feature.

**Comment by tlaurion** [comment]
A lot of discussions and ink happened on that subject before. I am not against merging this functionality and this work is an improvement on prior posture.

The only problem I see with this is that users are unaware of the consequences of those choices. For example, HID offering USB Security dongles will have QubesOS being setuped with holes punched under sys-usb if a USB Security dongle offering HID is present on second stage install phase, same for USB keyboards which will have final QubesOS installation opened to rubber ducky attacks.

Nothing we can really do for the past manufacturer choices to ditch PS2 and choose usb for convenience when times comes to support keyboard input.

But I would want more warnings there. For user knowledge of the risks they expose themselves, but also for future hardware makers to revisit that choice. We had lots of debates in the past, which librem key/HOTP withtout TPM was refused to be merged until at least firmware was fixed to not offer HID on newer firmware and Heads to seperate properly USB as transport and USB as input. 

This PR revisits that. What we could do:
- For boards requiring USB keyboards, warn every boot that USB rubber ducky attacks are part of user threat model.
- For board supporting USB keyboard optionally: force big warning that users are choosing convenience ofver security.

@JonathonHall-Purism thoughts?

**Comment by tlaurion** [comment]
Tested:
- [ ] x230-hotp-maximized_usb-kb

**Comment by JonathonHall-Purism** [comment]
:wave:  Sorry, this came out pretty long.  But since we have this discussion a lot, and I think it's important to have a general standard for consistency throughout Heads, I'm trying to approach it in a general sense and apply that to the situation at hand.

Take your time reviewing and I look forward to hearing your thoughts :+1: 

## In general

>But I would want more warnings there.

This is a discussion we have a lot.  I think it is time to think about some standards for this, so we can work toward consistent behavior throughout Heads.  Keep in mind that if we warn all the time, users will become accustomed to ignoring them, and it will be impossible to get their attention when something really important comes up.

Whenever we consider adding a warning, we need to weigh factors like:
* What are the consequences of the situation we might warn about?
* What can the user do about it?
* How much of the above does the user already know?

These are probably all ranges most of the time - consequences are a range of risks depending on the user's behavior, users will have a range of foreknowledge, etc.

All of those factors help us understand whether a warning will be perceived as useful or noise.  There may still be other ways to present information that would be perceived as "noise" (maybe we could provide a "report about the security of your configuration" or something?)

A bit aside, but this also should influence design at a much earlier phase.  We should design functionality as much as possible to reduce the need for warnings - but in this case I don't think there is much else we can do. (Maybe optionally dedicating a specific USB port for a USB keyboard in a later iteration?  This would prevent your compromised webcam from becoming a keyboard if it's possible, but would be difficult to explain without chassis markings on the ports.)

## Boards requiring USB keyboards

So what does that say about potentially warning during every boot on a board that requires a USB keyboard?

"What are the consequences of the situation we might warn about?"
* Qubes users: Compromising a USB device could lead to a system compromise that **would** be indicated by Heads' tamper indicators (whether tampering firmware or OS).
* Non-Qubes users: Less risk than Qubes users I think, since compromised USB devices can already gain a lot of access through the OS, though perhaps not quite as much as through controlling Heads.  For example, it could exfiltrate all the authentication tokens from your browser without even having to tamper anything.  Allowing it to control Heads as well means it could gain root/firmware-level access, but that would still be indicated by Heads' tamper indicators.
* There is no risk of an unindicated tamper here I think.  A compromised USB device still would not know your PIN to sign a tampered /boot, etc.

"What can the user do about it?"
* Buy another computer?  I don't think there is any alternative.  Disabling USB keyboards is not viable.  I don't think a computer with no keyboard control in firmware is viable even as a headless server (you have to be able to get in and boot to USB in case of a big problem).  So it is unlikely the user will take any action as a result of this warning.

"How much of the above does the user already know?"
* Qubes users probably know something about this already.  If you install Qubes yourself, it tells you about this, and you have to understand USB device isolation to some degree to use Qubes effectively.
* If you buy such a device with Heads and Qubes pre-installed, you might not know about this specific risk.  Some of those users probably have used Qubes before and might know about it, but some probably have not.  You still have to learn a bit about USB device isolation pretty quickly to use Qubes effectively.
* Non-Qubes users probably do not know about this.

I do not think that assessment justifies a warning, much less a noisy warning during every boot.  It could be useful information for a "security assessment" or similar information to report when requested.  Non-Qubes users are least likely to be aware of the risk, but are also the least impacted, since the bulk of the risk is already there in the OS.  While some Qubes users may not be aware of this risk, tamper indications still protect them, and there are practically no actions they can take about it.

## For board supporting USB keyboard optionally

For the sake of brevity I won't repeat all the above, but the balance is a little different here:
* You _can_ do something about it - disable USB keyboard support.  You are trading something off (potentially something significant if you have a physical issue using the built-in keyboard, or you modified the device, or it is broken, etc.), but it is a choice.
* However, it is mitigated again in that you had to turn this on, and we said why it is not on by default at that time (that's already there :wink: ).  You probably _are_ aware of this risk after that.

I think the balance here justifies informing the user of the warning when enabling the option, but not a persistent warning repeating that information during every boot.  And again, the risk is not a severe as it may initially seem, since tamper-evidence still applies.

**Comment by marmarek** [comment]
> Keep in mind that if we warn all the time, users will become accustomed to ignoring them, and it will be impossible to get their attention when something really important comes up.

This! There should be no situation when some intentional, conscious user decision results in a warning every boot.  There should be a warning when enabling the option, and changing it should be detected by tamper indicators (maybe even tell the user it's this setting changed triggered the tamper warning, if feasible). But for the system where USB keyboard is required, there is no point in showing warning every boot, because it isn't really actionable thing and will only make user less careful for more important warnings.

> (Maybe optionally dedicating a specific USB port for a USB keyboard in a later iteration? This would prevent your compromised webcam from becoming a keyboard if it's possible, but would be difficult to explain without chassis markings on the ports.)

Yes, this is a good idea for the future. For optional USB keyboard, enabling the feature could be also used to store ports, like: "Allow USB keyboard on the USB port(s) where it's currently connected". But that won't really work for the case of required USB keyboard, as that needs to work already before first user interaction.

**Review by tlaurion — CHANGES_REQUESTED** [review]

File: initrd/bin/config-gui.sh [hunk 1]
```diff
@@ -514,6 +524,34 @@ while true; do
         fi
       fi
     ;;
+    "K" )
+      if [ "$CONFIG_USER_USB_KEYBOARD" != "y" ]; then
+        if (whiptail --title 'Enable USB Keyboard?' \
+             --yesno "USB keyboards will be usable in $CONFIG_BRAND_NAME.
+                    \n\nIgnoring USB keyboards prevents a compromised USB device from controlling
```
[the remark this answers is not in the document]
Comment (marmarek) [reply initrd/bin/config-gui.sh:531]: Maybe rephrase "Ignoring ... prevents" to "Enabling ... allows", to avoid confusing positive/negative answer here. But otherwise yes, such warning when enabling is okay.

> probably removing x230-hotp-maximized_usb-kb 

Where is the config stored, in /boot? If so, for CI I may still need to have build time option then, because some tests involve wiping the whole disk clean, so the config won't survive (and then test cannot continue as USB input is no longer accepted).

[the remark this answers is not in the document]
Comment (JonathonHall-Purism) [reply initrd/bin/config-gui.sh:531]: That's a good thought, I think I like that wording better for the 'enable' prompt.  I'll give that a try.

I have that text in both the 'enable' and 'disable' prompts at the moment, I thought users would wonder why it was a setting at all (especially non-Qubes users).  But given that it's still off by default, maybe we don't need it in the 'disable' prompt, you would have seen the 'enable' text already.  I'll try only showing it on enable.

>Where is the config stored, in /boot?

It's the same as the other config menu settings, it's in config.user in flash.  So it will survive erasing /boot, but it will not survive a 'flash erasing settings' unless you insert a config.user in that flash image ahead of time (which works, I do it to offer preconfigured "basic" images and such).

[the remark this answers is not in the document]
Comment (marmarek) [reply initrd/bin/config-gui.sh:531]: > So it will survive erasing /boot

All good then, in that case x230-hotp-maximized_usb-kb could be dropped (at least for my needs).

[the remark this answers is not in the document]
Comment (tlaurion) [reply initrd/bin/config-gui.sh:531]: @marmarek as of this PR, x230-hotp-maximized_usb-kb will stay. But all other boards having native ps2 keyboard support will permit to enable usb keyboard support to be enabled from Config menu, save settings and warn that reboot is required (leading to TOTP/HOTP resealing and TPM DUK if enabled) since firmware is modified).

@JonathonHall-Purism Once message to user is changed on enablement, good to merge!

[the remark this answers is not in the document]
Comment (JonathonHall-Purism) [reply initrd/bin/config-gui.sh]: That change is up, thanks for patience as I had to prioritize a few other things before I could circle back to these PRs :bow:

**Review by tlaurion — APPROVED** [review]
